File: leetcode/0241-different-ways-to-add-parentheses/solution.cpp
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
        return solve(expression);
    }

private:
    unordered_map<string, vector<int>> memo;

    vector<int> solve(const string& expr) {
        auto it = memo.find(expr);
        if (it != memo.end()) {
            return it->second;
        }

        if (isNumber(expr)) {
            vector<int> single = {stoi(expr)};
            memo[expr] = single;
            return single;
        }

        vector<int> results;
        for (size_t i = 0; i < expr.size(); i++) {
            char ch = expr[i];
            if (ch == '+' || ch == '-' || ch == '*') {
                vector<int> left = solve(expr.substr(0, i));
                vector<int> right = solve(expr.substr(i + 1));
                for (int l : left) {
                    for (int r : right) {
                        if (ch == '+') results.push_back(l + r);
                        else if (ch == '-') results.push_back(l - r);
                        else results.push_back(l * r);
                    }
                }
            }
        }

        memo[expr] = results;
        return results;
    }

    bool isNumber(const string& s) {
        for (char c : s) {
            if (!isdigit(c)) return false;
        }
        return true;
    }
};

File: leetcode/0241-different-ways-to-add-parentheses/solution.py
from typing import List, Dict


class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        memo: Dict[str, List[int]] = {}

        def solve(expr: str) -> List[int]:
            if expr in memo:
                return memo[expr]

            if expr.isdigit():
                memo[expr] = [int(expr)]
                return memo[expr]

            results = []
            for i, ch in enumerate(expr):
                if ch in "+-*":
                    left = solve(expr[:i])
                    right = solve(expr[i + 1:])
                    for l in left:
                        for r in right:
                            if ch == "+":
                                results.append(l + r)
                            elif ch == "-":
                                results.append(l - r)
                            else:
                                results.append(l * r)

            memo[expr] = results
            return results

        return solve(expression)
